target/linux/ramips/image/mt7621_mikrotik.mk

```diff
@@ -13,14 +13,108 @@ define Build/append-dtb-elf
 		.appended_dtb=$(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb $@
 endef
 
+
+#######
+
+
+define linker-script-for-raw-dtb-appended-to-elf-preprocessor
+OUTPUT_ARCH(mips)
+SECTIONS {
+	/*
+	 * Preprocess the load address
+	 * -Ttext $ADDR is brittle:
+	 * https://lkml.kernel.org/lkml/20200413153453.zi4jvu3c4ul23e23@google.com/
+	 */
+	. = ZLOADADDR;
+	.text : {
+		startup = .;
+		*(.text)
+		*(.text.*)
+		*(.data)
+		*(.data.*)
+	}
+}
+endef
+
+# \n = $(newline), so that a mutli-line variable can be used
+# in shell through substituting $(newline) with '\n'
+# https://stackoverflow.com/a/5887751
+define newline
+
+
+endef
+
+# Replaces kernel-bin
+# requires ELF kernel
+# uses $@: target_rule_name example: /**/build_dir/target-*/linux-*/tmp/openwrt-*{,-initramfs}-kernel.bin
+# uses $<: first_prerequisite_name example: /**/build_dir/target-*/linux-*/$KERNEL_NAME{,-initramfs}
+# process:
+# copy the $KERNEL_NAME to $KDIR/tmp
+# convert to bytecode
+# append DTB
+# convert back to ELF (which requires data from the original ELF)
+# - ELF type example: elf32-tradlittlemips
+# - load address example: 0x80931000
+define Build/append-raw-dtb-to-elf
+	$(eval $@ : KERNEL_ELF_INPUT := $<)
+	$(eval $@ : KERNEL_OUTPUT := $@)
+	$(eval $@ : KERNEL_TEMP_UNLINKED := $(KERNEL_OUTPUT)_unlinked)
+	$(eval $@ : KERNEL_TEMP_BC := $(KERNEL_OUTPUT)_binary)
+	$(eval $@ : LINKER_SCRIPT := $(KERNEL_OUTPUT).lds)
+	$(eval $@ : LINKER_SCRIPT_UNPROCESSED := $(LINKER_SCRIPT).S)
+
+	$(eval $@ : ZLOADADDR := $(filter-out Entry point address:, \
+				   $(shell readelf -h $(KERNEL_ELF_INPUT) | \
+				     grep Entry)))
+	$(eval $@ : ELF_FORMAT := $(filter-out $(KERNEL_ELF_INPUT): file format, \
+				    $(shell $(TARGET_CROSS)objdump -h $(KERNEL_ELF_INPUT) | \
+				      grep 'file format' )))
+
+	$(eval $@ : TEMP_FILENAMES := $(KERNEL_TEMP_UNLINKED) \
+				      $(KERNEL_TEMP_BC) \
+				      $(LINKER_SCRIPT) \
+				      $(LINKER_SCRIPT_UNPROCESSED))
+
+	$(info KERNEL_OUTPUT=$(KERNEL_OUTPUT))
+	$(info KERNEL_ELF_INPUT=$(KERNEL_ELF_INPUT))
+	$(info ZLOADADDR=$(ZLOADADDR))
+	$(info ELF_FORMAT=$(ELF_FORMAT))
+
+	rm -f $(KERNEL_OUTPUT)
+	rm -f $(TEMP_FILENAMES)
+
+	$(TARGET_CROSS)objcopy -O binary \
+			       -R .reginfo -R .note -R .comment -R .mdebug -R .MIPS.abiflags -S \
+			       $(KERNEL_ELF_INPUT) $(KERNEL_TEMP_BC)
+	cat $(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb >> $(KERNEL_TEMP_BC)
+
+	$(TARGET_CROSS)ld -r -b binary --oformat="$(ELF_FORMAT)" \
+			  -o $(KERNEL_TEMP_UNLINKED) $(KERNEL_TEMP_BC)
+
+	printf '$(subst $(newline),\n,$(linker-script-for-raw-dtb-appended-to-elf-preprocessor))' \
+		  > $(LINKER_SCRIPT_UNPROCESSED)
+	$(TARGET_CROSS)cpp -P -Umips -U$(ARCH) -DZLOADADDR="$(ZLOADADDR)" \
+			   -D__ASSEMBLY__ -DLINKER_SCRIPT -o $(LINKER_SCRIPT) \
+			   $(LINKER_SCRIPT_UNPROCESSED)
+	$(TARGET_CROSS)ld -e startup -T $(LINKER_SCRIPT) -z max-page-size=4096 \
+			  -o $(KERNEL_OUTPUT) $(KERNEL_TEMP_UNLINKED)
+
+	rm -f $(TEMP_FILENAMES)
+endef
```

Comment:
Thanks @981213 and @f00b4r0!
The RAW_APPENDED_DTB ELF -> bytecode + append DTB -> ELF booted for me with this, and is easier than poking the kernel.

I threw the commit up atop the rest to get anyone's suggestions on my Makefile mess.
Where should this live? Under include/ if it also works for arm/ath79?
I will remove the subtarget, kernel patches, and trim it back.

Ccheers!